::tooltip pseudo element explainer (#1032)
MicrosoftEdge/MSEdgeExplainers: CSSTooltipPseudo/explainer.md

As the solutions to problems described in this document progress along the
standards-track, we will retain this document as an archive and use this section
to keep the community up-to-date with the most current standards venue and
content location of future work and discussions.

* This document status: **Active**
* Expected venue: [CSS Working Group](https://www.w3.org/Style/CSS/)
* Current version: this document

## Table of Contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Introduction](#introduction)
- [Goals](#goals)
- [Non-goals](#non-goals)
- [User research](#user-research)
- [Existing and upcoming tools available for custom tooltips](#existing-and-upcoming-tools-available-for-custom-tooltips)
  - [Use an existing library or component](#use-an-existing-library-or-component)
  - [Create a fully custom tooltip](#create-a-fully-custom-tooltip)
    - [Utilize `popover`, CSS anchor positioning, and/or `interesttarget`](#utilize-popover-css-anchor-positioning-andor-interesttarget)

> - [Current landscape of built-in tootlips](#current-landscape-of-built-in-tootlips)

Author: Yeah, I was shocked how many instances there were. Yikes :) Thanks for catching these

Tooltips are used by authors to provide additional contextul information to users
for a given interactive element on the screen. Tooltips are often presented as
a textbox that appears on hover (or potentially some other user interaction, like
keyboard focus, or long-press on a touchscreen device) on the element the
tooltip is associated with.

Authors can utilize the browser's built-in tooltips by setting the
[`title`](https://html.spec.whatwg.org/multipage/dom.html#attr-title)
attribute on an element of interest, or via the [`<title>`](
https://html.spec.whatwg.org/multipage/semantics.html#the-title-element)
element in SVG. However, the tooltip that appears as a result is not
customizable in any way by the author.

This lack of customizability has been a source of common frustration for authors
who want to keep a consistent look and feel across all of their UI.

This document explores a new `::tooltip` pseudo element that aims to provide authors
with a low effort mechanism to adjust styles for browser built-in tooltips without
having to create a fully custom solution, leaving positioning, input handling, and
accessibility for their tooltips up to the browser.

## Goals

> * Provide low effort way for authors to style core properties on browser built-in

Reviewer: "Provide a low effort..."

Author: Done

built-in tooltips (for example, color properties, font properties, etc.)
* This mechanism should handle accessibility, positioning, and input handling for
the author by default.
* Allow authors to override the default positioning behavior of the built-in
tooltips.
* Allow authors to change the timing of built-in tooltips.
* Improve accessibility of the `title` attribute, particularly for keyboard users.
* Guarantee security best practices, given that tooltips today can extend past
the bounds of the browser window.
* Styling built-in tooltips should produce consistent results across all browsers 
on the same OS, ensuring an interoperable experience for authors.

## Non-goals

* This proposal is not meant to support tooltips with arbitrary HTML content, which
can be covered by existing solutions, like `popover`, CSS Anchor positioning, and
the not-yet-available `interesttarget` attribute.

> * Provide a means to set a pointer on the `title` based tooltip. The solution should

Author: Good "point" :) Updated

be specified to allow for this functionality in the future, though.

## User research

Use cases in this explainer were collected from the discussion in issue
[8930](https://github.com/w3c/csswg-drafts/issues/8930). Additional inspiration
was drawn from discussions in issues
[9447](https://github.com/w3c/csswg-drafts/issues/9447), and OpenUI issue
[730](https://github.com/openui/open-ui/issues/730).

> There are also many articules on the web that walk authors through various

Author: Done

for tooltips on the web.

## Existing and upcoming tools available for custom tooltips

If an author would like to adjust the style of their tooltip today, they must resort
to one of the following alternative approaches to the built-in tooltip that they
get with the `title` attribute.

### Use an existing library or component

Authors can use an existing library or component for a fully customizable tooltip.
Some example libraries include:
  - https://fluent2.microsoft.design/components/web/react/core/tooltip/usage
  - https://www.opentip.org/
  - https://floating-ui.com/?utm_source=popper.js.org
  - https://www.npmjs.com/search?ranking=popularity&q=tooltip
  - Many more   

It is clear that there are many libraries and components for creating custom tooltips
that authors can use to solve their need for more customization.

If the library chosen is well architected to properly handle accessibility, this can
also be a convenient way to get such functionality "for free". However, relying on
external library to get accessibility right can be risky, which may limit the usability
if not done correctly.

The other downside of this approach is that it requires heavy use of scripting,

> which can be slower, and shouldn't be necessary if all you want to do is

Reviewer: "You want" seems a little vague here, how about "if the author simply wants to"

Author: Done

adjust some basic styles.

### Create a fully custom tooltip

Authors can also choose to create their own custom tooltip or component, which
requires their own input management, proper accessibility support and proper
handling of the tooltip positioning, which can be complex.

This does allow authors to acheive their goal of full cusomizability, but it often
requires heavy use of scripting, or use of `:before` or `:after` pseudos, which
can make getting accessibility correct quite difficult.

As such, this solution requires a lot of work to get right, especially if all they
want to do is adjust a few simple styles on the built-in tooltip.

#### Utilize `popover`, CSS anchor positioning, and/or `interesttarget`

Although creating a custom tooltip can be cumbersome for an author who is looking
to just adjust a few simple styles, they do have some new (and upcoming) web
APIs, that can help them in accomplishing a fully customizable tooltip more
easily:
  - The [`popover`](https://html.spec.whatwg.org/multipage/popover.html#dom-popover) 

>     attribute, which is baseline across browsers as of April 2024.

Author: Good idea - link added

https://drafts.csswg.org/css-ui/#propdef-appearance), similar to
what was done for [customizable select elements](
https://developer.mozilla.org/en-US/docs/Learn_web_development/Extensions/Forms/Customizable_select),
called `base-tooltip`, that an author can use within the `::tooltip`
pseudo element to trigger the new base default styles and
customizability.

For example, in order to set the `color` of your tooltip to `red`,
one could apply the following styles:

```css
::tooltip {
  appearance: base-tooltip;
  color: red;
}
```

The downside of this is that it requires explicit opt-in, rather than
only needing to set the `color` to `red`, but would allow an author to
trigger the new base tooltip styles without having to set any styles.
This triggering mechanism remains an open question and is welcome
to additional suggestions and input from the wider community.

### How to set the `::tooltip` content?

There has been debate in the [most recent proposal](
https://lists.w3.org/Archives/Public/www-style/2000Apr/0014.html)
for `::tooltip` for how best to set the content of the `::tooltip`
pseudo element.

The content itself will be pulled from the `title` attribute, or in

> the case of SVG, the `<title>` element. The initial proposal is the

Author: Done

as the proposal to define the position via CSS Anchor Positioning
(which was included as an option in Lea's original proposal, as
well).

```css
/* UA styles */
::tooltip {
  /* content: normal; is implied */
  color: InfoText;
  background: InfoBackground;
  font-size: .8rem;
  box-shadow: .1rem .1rem .2rem rgb(0 0 0 / .2);
  position: absolute;
  position-area: top right;
  position-try-fallbacks: top left, top center, bottom right, bottom left, bottom center, center right, center left;
  transition: 0s 3s visibility;
  
  @starting-style {
    visibility: hidden;
  }
}
```

These proposed set of initial UA styles is a starting point and is
open to further discussion.

To see some of these styles in action, check out the [CodePen demo](
https://codepen.io/alisonmaher/pen/MYYxpJZ) utilizing `interesttarget`
to render the tooltip. You'll note that this requires some changes
to the styles above (notably, this doesn't include the defined
`transition` and sets `margin` to `0` to account for the default
`popover` styles). Since this is based on `interesttarget`, at the
time of writing, this demo is required to be tested in a Chromium-based

> browser with Web Platform Experimental Flags Enabled.

Author: Updated, although the weird thing with this is that in Edge it is edge:// and this is an Edge explainer. Wasn't sure what link to include so left it blank, but this is probably (unfortunately) the safest link :)

> browser with Web Platform Experimental Flags Enabled.

Reviewer 2: Yeah that's fair. The `chrome` one will work in Edge too at least.

> browser with Web Platform Experimental Flags Enabled.

Author: Oh I didn't know that worked in Edge, good to know. This makes me feel better then. Resolved!

}
```

Whether authors should be able to change the `content` of these tooltips
is still an open question. If we decide that is not allowed, this would
require a different mechanism (like utilizing `visibility`) if we want
to support this use case for authors.

### Scenario 5: Providing custom positioning to a tooltip

Given that this proposal uses CSS Anchor Positioning to define the base
positions of the `::tooltip` pseudo element, this would open up an
opportunity for authors to adjust these styles, similar to the example
below:

```css
::tooltip {
  appearance: base-tooltip;
  position-area: start span-all;
}
```

## Accessibility Considerations

According to the [WCAG](
https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html),
“Content which can be triggered via pointer hover should also be able to be triggered by keyboard focus.” As such, tooltips should also be keyboard
accessible, and per guidance should also be dismissible.

### What is the current accessibility experience of tooltips across browsers?

Unlike [`interesttarget`](
https://open-ui.org/components/interest-invokers.explainer/#why-is-interesttarget-not-unlimited-like-title-is),

> the `title` attribute is allowed on any element, even if it isn't

Author: Done

interactable. This can pose problems for accessiblity, because if the
element isn't focusable, it means that the associated tooltip is not
keyboard accessible.

On top of this, no browser, outside of Microsoft Edge, exposes the
built-in tooltip on focus, even if the associated element is focusable.
This means that browser built-in tooltips are not accessible
out-of-the-box, which is one of the goals of this proposal.

### Proposed improvements

> To improve the current situation, the `::tootlip` element cannot be

Author: Thanks for the catch! And was done an embarrassingly large number of times, so updated them all

triggered or styled if the element associated with the tooltip is
not focusable.

> As part of this effort, browser vendors should also work on triggering

Author: Yeah that is the idea, although engineering work here will be delayed due to reshuffling of work. But this would be something that should be done as part of this feature work

their built-in tooltips on focus when the element associated with the
tooltip is focusable, and utilize the escape key to trigger tooltip
dismissal.

> One consideration for improving this further is to force existing

Author: Right, updated the text to clarify

`title`-based tooltips to have a `tabindex` of `0` if it is not
already focusable. However, this is not ideal, as it could lead
a container to become focusable that was not before, leading to
undesirable focus changes if a user incorrectly applies focus just
outside of a target element within that container.

An alternative for browser vendors to consider is including a user
setting that forces a `tabindex` of `0` in such cases, allowing the
user more control over the resulting behavior as they see fit.

Creating a more accessible tooltip experience for users is one of
the key goals of this proposal, so additional ideas from the
community that should be considered to help in further improving

> the current situation is very much welcome.

Author: Done

- [Security Considerations](#security-considerations)
- [Privacy Considerations](#privacy-considerations)
- [Open questions](#open-questions)
- [Future ideas](#future-ideas)
  - [Tootlip pointer/arrow](#tootlip-pointerarrow)
  - [Customizing `::tooltip` user interactions](#customizing-tooltip-user-interactions)
  - [CSS Anchor Positioning Tooltip Defaults](#css-anchor-positioning-tooltip-defaults)
- [Considered alternatives](#considered-alternatives)
  - [Alternative 1: The `interesttarget` attribute](#alternative-1-the-interesttarget-attribute)
- [References & acknowledgements](#references--acknowledgements)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Introduction

Tooltips are used by authors to provide additional contextul information to users
for a given interactive element on the screen. Tooltips are often presented as
a textbox that appears on hover (or potentially some other user interaction, like
keyboard focus, or long-press on a touchscreen device) on the element the
tooltip is associated with.

Authors can utilize the browser's built-in tooltips by setting the
[`title`](https://html.spec.whatwg.org/multipage/dom.html#attr-title)

> attribute on an element of interest. However, the tooltip that appears as a

Reviewer: Might also be worth calling out the SVG `<title>` element, which works similarly: https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Element/title

Author: Good call out. Forgot to add it at this location. Updated!

result is not customizable in any way by the author.

This lack of customizability has been a source of common frustration for authors
who want to keep a consistent look and feel across all of their UI.

This document explores a new `::tooltip` pseudo element that aims to provide authors
with a low effort mechanism to adjust styles for browser built-in tooltips without
having to create a fully custom solution, leaving positioning, input handling, and
accessibility for their tooltips up to the browser.

## Goals

* Provide low effort way for authors to style core properties on browser built-in
tooltips (for example, color properties, font properties, etc.)
* This mechanism should handle accessibility, positioning, and input handling for
the author by default.
* Allow authors to override the default positioning behavior of the built-in
tooltips.
* Allow authors to change the timing of built-in tooltips.
* Improve accessibility of the `title` attribute, particularly for keyboard users.

> * Guarentee security best practices, given that tooltips today can extend past

Reviewer: Typo:  Guarentee->Guarantee

> the bounds of the browser Window.

Reviewer: Window doesn't need to be capitalized

Author: Done

* Styled tooltips should produce consistent results across all platforms, ensuring

> an interoperable experience for authors.

Reviewer: Might be worth noting two things here: 1) this is not possible today, and 2) this proposal will also allow for interoperable results between browsers on the same OS, but maybe "platforms" is generic enough to cover OS and browser.

Author: Since there is no way to style the default tooltips today, I don't know how best to note the first one. But I've updated the wording to add more clarity that this is specifically for styling of built-in tooltips, and also clarified interop across browsers on the same OS (although technically the goal is across OS, as well, for the styled case)

## Non-goals

> * This proposal is not meant to solve for tooltips with arbitrary HTML content, which

Reviewer: "...is not meant to solve for tooltips with..." -> "...is not meant to support tooltips with..."

Author: Done

can be covered by existing solutions, like `popover`, CSS Anchor positioning, and
the not-yet-available `interesttarget` attribute.
* Provide a means to set a pointer on the `title` based tooltip. The solution should
be specified to allow for this functionality in the future, though.

## User research

Use cases in this explainer were collected from the discussion in issue
[8930](https://github.com/w3c/csswg-drafts/issues/8930). Additional inspiration
was drawn from discussions in issues
[9447](https://github.com/w3c/csswg-drafts/issues/9447), and OpenUI issue
[730](https://github.com/openui/open-ui/issues/730).

There are also many articules on the web that walk authors through various
options for creating custom tooltips using tools available to them today.
One such [article](https://blog.replaybird.com/css-tooltip-examples/)
walks through different options, along with various examples and use cases
for tooltips on the web.

## Existing and upcoming tools available for custom tooltips

If an author would like to adjust the style of their tooltip today, they must resort

> to the one of the following alternative approaches to the built-in tooltip that they

Reviewer: "to the one of the" -> "to one of the"

get with the `title` attribute.

### Use an existing library or component

Authors can use an existing library or component for a fully customizable tooltip.
Some example libraries include:
  - https://fluent2.microsoft.design/components/web/react/core/tooltip/usage
  - https://www.opentip.org/
  - https://floating-ui.com/?utm_source=popper.js.org
  - https://www.npmjs.com/search?ranking=popularity&q=tooltip
  - Many more   

It is clear that there are many libraries and components for creating custom tooltips
that authors can use to solve their need for more customization.

> If the library chosen if well architected to properly handle accessibility, this can

Reviewer: "chosen if well architected" -> "chosen is well architected"

Author: Done

also be a convenient way to get such functionality "for free". However, relying on
external library to get accessibility right can be risky, which may limit the usability
if not done correctly.

> The other downside of this approach is that it requires heavy use of JS, which can be

Reviewer: Might want to spell out "Javascript", or just say "scripting"

Author: Done

slower, and shouldn't be necessary if all you want to do is adjust some basic styles.

### Create a fully custom tooltip

Authors can also choose to create their own custom tooltip or component, which
requires their own input management, proper accessibility support and proper
handling of the tooltip positioning, which can be complex.

This does allow authors to acheive their goal of full cusomizability, but it often
requires heavy use of JS, or use of `:before` or `:after` pseudos, which can make
getting accessibility correct quite difficult.

As such, this solution requires a lot of work to get right, especially if all they
want to do is adjust a few simple styles on the built-in tooltip.

#### Utilize `popover`, CSS anchor positioning, and/or `interesttarget`

Although creating a custom tooltip can be cumbersome for an author who is looking
to just adjust a few simple styles, they do have some new web APIs, and APIs that

> are in the works, that can help them in accomplishing a fully customizable tooltip

Reviewer: "they do have some new web APIs, and APIs that are in the works" - maybe say "they do have some new (and upcoming) web APIs""

Author: Done

attribute.
_____________________________

All of these options do allow authors to create more custom tooltip experiences
than what is currently available through the browser's built-in option using the
`title` attribute, but these options often require more work than should be
necessary for a simple style adjustment.

## Current landscape of built-in tootlips

The rendering of tooltips via the `title` attribute is up to each UA. Below
is a brief survey of how some of the top browser engines render built-in
tooltips today, in no particular order.

### Simple tooltips across browsers

In this section is a comparison of basic tooltip rendering across each
browser engine using the code example below.

```html
<button title="This is a button!">Button Example</button> 
```

#### Chromium

In Chromium, the tooltip is rendered above all web content, and the position
of the tooltip is dependent on the position of the mouse and the element it
is anchored to.

The style of the tooltip that is rendered is very simple: a white or black
background with black or white text and a matching border.

On Windows the simple test case above renders as follows:

> <p align="center">

Reviewer: These images don't show up for me. Is this just something to do with the PR, and they'll work once it's landed?

Should they be using markdown image syntax instead of HTML? https://www.digitalocean.com/community/tutorials/markdown-markdown-images

Also, these should have alt text 🤓

Author: I had looked at the Gap Decorations explainer as an example for images. But updated to the markdown version (which also means I wasn't able to center them, which isn't a problem).

Also added alt text.

The images show up for me locally, so I think this is an issue with the Github tooling. But this happened with other explainers I reviewed, so they should show up correctly once this lands.

Many have proposed a pseudo for styling browser tooltips over the years,
with the [first proposal](
https://lists.w3.org/Archives/Public/www-style/2000Apr/0014.html)
originating around April 2000, and the most [recent proposal](
https://github.com/w3c/csswg-drafts/issues/8930#issuecomment-1581848840)
from Lea Verou in 2023. 

The proposal in this document takes close inspiration from these, by
introducing a new pseudo element, called `::tooltip`, that can be
used by authors to style the browser tooltip that is displayed when
the `title` attribute is set on an HTML element.

The `::tooltip` pseudo element will have default UA styles that will
be required to be applied when an author adjusts tooltip styles
within a `::tooltip` pseudo element selector.

The idea behind this proposal is that authors would be able to
create simple text-based tooltips, and adjust the styles of color
properties, shadowing, border properties, font properties, etc
to better match the look and feel of their site, leaving
positioning, input handling, and accessibility up to the browser.

### How to trigger `::tooltip` default styles?

Currently, browser default tooltips are triggered by the `title`

> attribute being set, or `<title>` being set within an SVG element.

Reviewer: "or `<title>` being set within an SVG element." Maybe phrase this as "or the presence of a `<title>` child node for SVG elements."

We could have the default `::tooltip` apply in all cases by default,

> but this could be a breaking change for browsers. Instead, the UA

Reviewer: Is this really a breaking change if all OS's and browsers are currently different? Seems like it's safe to change.

Author: It might not be completely breaking (but potentially for no longer allowing tooltips to escape window bounds).

We will likely look into unifying things in Chromium, but this would be something that would be done after this work, and shouldn't be a requirement for UAs.

Updated the working slightly to say it may be a breaking change on some browsers, since it is unclear if it will or won't

should be able to style tooltips in their own chosen way, unless
`::tooltip` is triggered in some way.

> Authors may also want to trigger this path without triggering any

Reviewer: Interesting, so the legacy behavior will still exist for most sites, unless they use a  `::tooltip` selector?

Author: Yeah the idea is that we allow UAs to keep whatever styles they want outside of this case.

This would likely be preferable for Safari, since they will likely want to continue to match the OS in this case.

This could also allow us to update our default styling on Windows to better match Windows, which is an ask from one of our customers (CC @aluhrs13 who likely remembers who that is)

for `::tooltip` for how best to set the content of the `::tooltip`
pseudo element.

The content itself will be pulled from the `title` attribute, or in
the case of SVG, the `<title>` element. The initial proposal is to
utilize the `content` property for applying the content to the
`::tooltip` pseudo element itself.

The initial suggestion by Lea Verou in the CSSWG proposal was to
set this via the default stylesheet in the following manner:

```css
::tooltip {
  content: attr(title);
  ...
}
```

There were some concerns that this wouldn't work well for SVGs,
since SVGs use `<title>` instead of the `title` attribute for
built-in tooltips. [@bleper](https://github.com/bleper) put
together a [potential proposal](
https://github.com/w3c/csswg-drafts/issues/8930#issuecomment-1581848840)
for how to address this issue.

However, a simpler approach may be to define what `normal` means
uniquely for `::tooltip` such that `normal` is equivalent to
`content: attr(title)` in all cases, but SVG, where in SVG,
`normal` is equivalent to the content of the `<title>` element.

This item remains an open question and is welcome to input from
the wider community.

> ### Positioning for `::tooltip`

Reviewer 3: Should there be something written about tooltip dimensions? For example, can the web developer control the dimensions? Can the dimensions be modified via script, animation etc. after creation? Are there constraints other than 'limited to the browser window size'?

Author: This is a good callout. I had considered adding something about this at one point and completely forgot to add a section on it, so appreciate you bringing it up.

For now, I've defined on using default sizing constraints, which means that the tooltip will be sized based on its text content (and padding properties, etc). I also played with the demo in https://codepen.io/alisonmaher/pen/MYYxpJZ and with longer tooltips, the size is constrained by the viewport in the inline direction, but might overflow in the block direction.

I've included this info in a new section (called ::tooltip sizing), and leave an open question for if this should be adjustable by the author (I don't see a strong use case for this since the only content can be text, but I also wouldn't object to adding such properties if we discover a good reason from developers).

I was going to add an open question on if we should set a non-default sizing constraint on the tooltip, but figured that is covered by the open question on "what are the right default UA styles?", but feel free to let me know if you think it is worth calling that out explicitly, as well.

### Positioning for `::tooltip`

There has been discussion in the CSSWG around whether the positioning
of `::tooltip` should be UA magic, or if the author should be able to
customize this. The main concern being around security, given that
tootlips can escape the bounds of the browser window today (see
[Security Considerations](#security-considerations) for more details).

Given the proposal to not allow tooltips styled through the
`::tooltip` psuedo element to escape the bounds of the windows (see
[Security Considerations](#security-considerations) for more details),
this proposal also includes the ability for authors to customize the
tooltip positioning.

As such, this proposal will utilize [CSS Anchor Positioning](
https://drafts.csswg.org/css-anchor-position-1/) to define the default
position of the `::tooltip` pseudo element, which will allow authors
to override this in the way that they see fit.

> As a result, this mean that the element associated with the `title`

Reviewer: "this means"

Author: Done

  transition: 0s 3s visibility;
  
  @starting-style {
    visibility: hidden;
  }
}
```

These proposed set of initial UA styles is a starting point and is
open to further discussion.

To see some of these styles in action, check out the [CodePen demo](
https://codepen.io/alisonmaher/pen/MYYxpJZ) utilizing `interesttarget`
to render the tooltip. You'll note that this requires some changes
to the styles above (notably, this doesn't include the defined
`transition` and sets `margin` to `0` to account for the default
`popover` styles). Since this is based on `interesttarget`, at the
time of writing, this demo is required to be tested in a Chromium-based
browser with Web Platform Experimental Flags Enabled.

### What styles can be applied to `::tooltip`?

Not all styles should or need to be applied within the context
of a simple text-based tooltip. For example, we likely don't
want tooltips with images, lists, scrollbars, or non-default
`display` types or `z-index`. As such, there should be a limited list of styles
that are valid within a `::tooltip` pseudo element; that list of

> which can be expanded on in the future.

Reviewer: "that list of which can be expanded on in the future." -> "that list can be expanded on in the future."

Author: Done

There is an open [CSSWG issue](
https://github.com/w3c/csswg-drafts/issues/9447) to discuss what
properties should be styleable within a `::tooltip` pseudo element.
As such, the complete list of properties is up for discussion, but
as the proposal notes, the core set of properties should include:
- `content` for setting the text displayed in the tooltip (whether
authors should be allowed to change this is up for debate)
- `appearance` as a means to trigger new default tooltip styles
- All font properties, `text-transform`, `letter-spacing`,
`word-spacing`, and the i18n text properties
- `background`, `color`, `box-shadow`, `text-shadow`,
`border-radius`, `border`, `padding`, `opacity` (including their
longhands)
- `color-scheme` and `forced-color-adjust` (these two were not
included in the initial [CSSWG proposal](
https://github.com/w3c/csswg-drafts/issues/9447), but would likely
be helpful for authors to guarantee consistent theming capabilites
with the rest of their page.)
- Transitions

> - `tranform`, `transform-origin`, `rotate`, `scale`, `translate`

Reviewer 3: Nit: 'tranform' -> 'transform'

Author: Done

There is no existing way to set an [accessible description](
https://developer.mozilla.org/en-US/docs/Glossary/Accessible_description#accessible_description_computation) for most
elements without using ARIA or without a tooltip being shown (via
the `title` attribute).

We can consider making this possible through the `::tootlip` proposal
utilizing `content: none` as follows:

```css
::tooltip {
  appearance: base-tooltip;
  content: none;
}
```

Whether authors should be able to change the `content` of these tooltips
is still an open question. If we decide that is not allowed, this would
require a different mechanism (like utilizing `visibility`) if we want
to support this use case for authors.

### Scenario 5: Providing custom positioning to a tooltip

Given that this proposal uses CSS Anchor Positioning to define the base
positions of the `::tooltip` pseudo element, this would open up an
opportunity for authors to adjust these styles, similar to the example
below:

```css
::tooltip {
  appearance: base-tooltip;
  position-area: start span-all;
}
```

## Accessibility Considerations

According to the WCAG, “Content which can be triggered via pointer hover

> should also be able to be triggered by keyboard focus.” (Resource). As

Reviewer: Is (Resource) supposed to be a link?

such, tooltips should also be keyboard accessible, and per guidance should
also be dismissible.

### What is the current accessibility experience of tooltips across browsers?

Unlike [`interesttarget`](
https://open-ui.org/components/interest-invokers.explainer/#why-is-interesttarget-not-unlimited-like-title-is),
the `title` attribute is allowed on any element, even if it isn't
interactable. This can pose problems for accessiblity, because if the
element isn't focusable, it means that the associated tooltip is not
keyboard accessible.

> On top of this, no browser, outside of Microsoft Edge, exposes to

Reviewer: "exposes to" -> "exposes the"

Author: Done

the key goals of this proposal, so additional ideas from the
community that should be considered to help in further improving
the current situation is very much welcome.

### Zoom

Built-in tooltips today do not adhere to user preferred zoom levels,
making them less accessible than the content they are associated with.

Making these tooltips stylable through the new `::tooltip` pseudo
element should allow for a more accessible experience in such
scenarios.

## Security Considerations

As noted in the section on the [current landscape of built-in
tootlips](#current-landscape-of-built-in-tootlips), browser built-in
tooltips are currently allowed to escpape the bounds of the browser
window.

For tooltips styled via the `::tooltip` pseudo element, UAs **must** instead
stay within the bounds of the web page. This would avoid [security concerns](
https://textslashplain.com/2017/01/14/the-line-of-death/) when rendering
outside the bounds of the browser window, which becomes exacerbated when
styleability is involved, because this can more easily trick users into
clicking on a spot on the screen that could lead to a security vulnerability.

> This behavior would be limiting for Iframes in particular but would be

Reviewer: I was curious about the iframe behavior - with this proposal, tooltips would be limited to the content area of the iframe, right? Might be worth spelling out here.

Author: Added more details, thanks for the callout

- Should `::tooltip` be based on `interesttarget` or should that
detail be left up to the UA?
- Should we consider another method for triggering `::tooltip` and its
base styles than the `appearance` property?
- Is `content` the best mechanism for setting the tooltip text from
`title` within the `::tooltip` psuedo element?
- Should authors be allowed to change the `content` of the `::tooltip`?
- Should authors be able to change the positioning of a `::tooltip` or
should this be magic left up to the UA?
- What are the right default UA styles for `::tooltip`?
- Does it make sense for `::tooltip` to be a [Tree-Abiding
Pseudo-element](https://drafts.csswg.org/css-pseudo/#treelike)?
- How should `::tootltip` interact with `interesttarget` and any other
`popover` on the same element?
- Are there other ways we improve the current accessibility landscape
of the `title` attribute?
- Should the text within `::tooltip` be selectable?
- How should `::tooltip` be represented in the DOM tree?
- Should we use `visibility` for showing the `::tooltip` transition,
or [`position-visibility`](
https://drafts.csswg.org/css-anchor-position-1/#position-visibility)?

> If `position-visibility`, we may want to consider adding a new value

Reviewer: I was going to raise another issue - what happens with `pointer-events: none`, but it looks like that already disables the native tooltips in Chromium and Firefox, so this wouldn't change anything in that regard (!)

Author: Ack, thanks for the callout, glad to hear this shouldn't be a problem!

Some example libraries include:
  - https://fluent2.microsoft.design/components/web/react/core/tooltip/usage
  - https://www.opentip.org/
  - https://floating-ui.com/?utm_source=popper.js.org
  - https://www.npmjs.com/search?ranking=popularity&q=tooltip
  - Many more   

It is clear that there are many libraries and components for creating custom tooltips
that authors can use to solve their need for more customization.

If the library chosen is well architected to properly handle accessibility, this can
also be a convenient way to get such functionality "for free". However, relying on
external library to get accessibility right can be risky, which may limit the usability
if not done correctly.

The other downside of this approach is that it requires heavy use of scripting,
which can be slower, and shouldn't be necessary if the author simply wants to
adjust some basic styles.

### Create a fully custom tooltip

Authors can also choose to create their own custom tooltip or component, which
requires their own input management, proper accessibility support and proper
handling of the tooltip positioning, which can be complex.

> This does allow authors to acheive their goal of full cusomizability, but it often

Reviewer 4: ```suggestion
This does allow authors to achieve their goal of full customizability, but it often
```

Author: Done

> requires heavy use of scripting, or use of `:before` or `:after` pseudos, which

Reviewer 4: ```suggestion
requires heavy use of scripting, or use of `::before` or `::after` pseudos, which
```

(Pseudo-classes such as `:hover` start with a single colon. Pseudo-elements such as `::before` start with a double colon.)

Author: Whoops, thanks for the catch. Updated!

On Windows the simple test case above renders as follows:

![Screenshot of tooltip in Gecko on Windows](
images/simple-button-firefox-windows.png)

On Linux, the style of the tooltip defaults to a darker color scheme,
with no border, and slightly more padding:

![Screenshot of tooltip in Gecko on Linux](
images/simple-button-firefox-linux.png)

On MacOS, Gecko matches the OS tooltip styles, producing the following
result:

![Screenshot of tooltip in Gecko on Mac](
images/simple-button-firefox-mac.png)

On mobile, tooltips are not activatable in Gecko.

#### Webkit

In Webkit, the tooltip is similarly rendered above all web content, and the
position of the tooltip is dependent on the position of the mouse and the
element it is anchored to.

The style of the tooltip that is rendered is in accordance with the platform,
similarly to Chromium and Firefox on MacOS.

![Screenshot of tooltip in Webkit on Mac](
images/simple-button-webkit-mac.png)

On mobile, tooltips are not activatable in Webkit.

### Longer tooltips across browsers

Below is an overview of how each major browser engine handles longer tooltip text,
in no particular order.

#### Chromium

> In Chromium, built-in tooltips add elipses to the tooltip text if the text

Reviewer 4: ```suggestion
In Chromium, built-in tooltips add ellipses to the tooltip text if the text
```

Author: Done

is greater than or equal to 1025 characters. The tooltip is also allowed to
escape the bounds of the window. For example:

> ![Screenshot of a long tooltip with an elipses in Chromium on Windows](

Reviewer 4: ```suggestion
![Screenshot of a long tooltip with an ellipsis in Chromium on Windows](
```

Author: Done

images/out-of-bounds-chromium-windows.png)

#### Gecko

In Gecko, I was *not* able to find a limit for built-in tooltip length for

> which an elipses was added, like it does in Chromium. However, tooltips

Reviewer 4: ```suggestion
which an ellipsis was added, like it does in Chromium. However, tooltips
```

Author: Done

are also able to escape the bounds of the window in Gecko, but the tooltip
width stays more constrained than in Chromium. For example:

![Screenshot of a long tooltip in Gecko on Windows](
images/out-of-bounds-firefox-windows.png)

#### Webkit

In Webkit, I was *not* able to find a limit for built-in tooltip length for

> which an elipses was added, like it does in Chromium. However, tooltips

Reviewer 4: ```suggestion
which an ellipsis was added, like it does in Chromium. However, tooltips
```

Author: Done

create simple text-based tooltips, and adjust the styles of color
properties, shadowing, border properties, font properties, etc
to better match the look and feel of their site, leaving
positioning, input handling, and accessibility up to the browser.

### How to trigger `::tooltip` default styles?

Currently, browser default tooltips are triggered by the `title`
attribute being set, or by the presence of the `<title>` element for
SVG elements.

We could have the default `::tooltip` apply in all cases by default,
but this may be a breaking change for some browsers. Instead, the UA
should be able to style tooltips in their own chosen way, unless
`::tooltip` is triggered in some way.

Authors may also want to trigger this path without triggering any
styles beyond those that are default. As such, there requires
some mechanism for authors to trigger the new `::tooltip`-based
default.

An initial suggestion would be to add a new value to [`appearance`](
https://drafts.csswg.org/css-ui/#propdef-appearance), similar to
what was done for [customizable select elements](
https://developer.mozilla.org/en-US/docs/Learn_web_development/Extensions/Forms/Customizable_select),

> called `base-tooltip`, that an author can use within the `::tooltip`

Reviewer 4: Why `base-tooltip` rather than `base` like we have in [CSS Forms 1](https://www.w3.org/TR/css-forms-1)?

Author: I was looking at the current definition of `appearance` which was why I added a new value for this, but that would make sense to me. Updated to use `base` and updated all example code using `base-tooltip` to `base`

the wider community.

### Positioning for `::tooltip`

There has been discussion in the CSSWG around whether the positioning
of `::tooltip` should be UA magic, or if the author should be able to
customize this. The main concern being around security, given that
tooltips can escape the bounds of the browser window today (see
[Security Considerations](#security-considerations) for more details).

Given the proposal to not allow tooltips styled through the
`::tooltip` psuedo element to escape the bounds of the windows (see
[Security Considerations](#security-considerations) for more details),
this proposal also includes the ability for authors to customize the
tooltip positioning.

As such, this proposal will utilize [CSS Anchor Positioning](
https://drafts.csswg.org/css-anchor-position-1/) to define the default
position of the `::tooltip` pseudo element, which will allow authors
to override this in the way that they see fit.

As a result, this means that the element associated with the `title`
attribute (or `<title>` element in the case of SVG) will be an implied
anchor for the rendered tooltip.

### Default styles for `::tooltip`

> When the new default `::tooltip` styles are tiggered via the

Reviewer 4: ```suggestion
When the new default `::tooltip` styles are triggered via the
```

Author: Done

`appearance` property, a set of default styles would be applied
by the UA.

These styles were mostly pulled directly from Lea Verou's initial
[proposal](
https://github.com/w3c/csswg-drafts/issues/8930) with a couple of
minor edits based on the new proposal for an updated definition of
`content: normal` within the `::tooltip` pseudo element, as well
as the proposal to define the position via CSS Anchor Positioning
(which was included as an option in Lea's original proposal, as
well).

```css
/* UA styles */
::tooltip {
  /* content: normal; is implied */
  color: InfoText;
  background: InfoBackground;
  font-size: .8rem;
  box-shadow: .1rem .1rem .2rem rgb(0 0 0 / .2);
  position: absolute;
  position-area: top right;
  position-try-fallbacks: top left, top center, bottom right, bottom left, bottom center, center right, center left;
  transition: 0s 3s visibility;
  
  @starting-style {

>     visibility: hidden;

Reviewer 4: It's a little odd that `visibility` is included among the UA default styles when it's not listed among the properties that can be applied to `::tooltip` in the next section. One might read these default styles as suggesting that an author could trigger a tooltip to be shown by doing something like:

```css
.myelement::tooltip {
  visibility: visible;
}
```

To be clear, I think we should disallow author control of visibility in this manner and leave it entirely up to the UA. If we enabled visibility control like in the example above, we would need to calculate styles for the `::tooltip` pseudo on every element just to check for visibility, which would be a big perf hit.

edit: Reading further I see there's a use case for adjusting the timing of showing/hiding the tooltip, but I'll leave this comment open for consideration of the point about style calc performance.

Author: Yeah agreed we don't want an author to play with it, but they might want to adjust the timing as you mentioned. Would we need to include `visibility` in the list of adjustable styles for an author to be able to adjust the `transition` tied to `visibility`?

Either way, I think `appearance: base` being the thing that triggers the base tooltip styles, as opposed to `visibility` on its own would remain in tact, which I think would avoid the performance concern?

There is also an open question on the right property to include in the transition (i.e. should it be `position-visibility` instead, which may require an additional value to `position-visibility`, but perhaps there is another option)?

We could also leave this as a future idea to let authors adjust this timing, which would leave the transition bit up to the UA as opposed to part of the UA stylesheet

Author: Although, I guess is that problem that an author could change the value of `appearance` via script, which means that we would need to store the ::tooltip styles in some way to ensure that works as expected?

Author: Chatted offline and got clarity on what was meant here. I've updated the note under Scenario 3 to better clarify why `visibility` isn't allowed to be adjusted by authors. Lmk if I missed anything.

https://github.com/w3c/csswg-drafts/issues/9447), but would likely
be helpful for authors to guarantee consistent theming capabilites
with the rest of their page.)
- Transitions
- `tranform`, `transform-origin`, `rotate`, `scale`, `translate`
- Custom properties
- Given that the proposal suggests using anchor positioning for
default styling of the position of `::tooltip`, we should also
include [CSS Anchor Positioning](
https://drafts.csswg.org/css-anchor-position-1/) properties
for author customization

The list above should handle the majority of styling use cases
for authors, however the [current proposal](
https://github.com/w3c/csswg-drafts/issues/9447) also notes that
the following properties would be nice to have, as well:
- `clip-path`
- Filters and `backdrop-filter`
- `animation`
- `mix-blend-mode`
- `border-image`
- `outline`
- `text-decoration`
- `fill` and `stroke`

As noted, the first list of valid properties is still up for
debate, so any thoughts or input are welcome as we continue to
explore the [open issue](
https://github.com/w3c/csswg-drafts/issues/9447).

### `::tooltip` style inheritance

To allow `::tooltip` to inherit styles from the rest of the page,

> it should be grouped witin the list of [Tree-Abiding

Reviewer 4: ```suggestion
it should be grouped within the list of [Tree-Abiding
```

Author: Done

Pseudo-elements](https://drafts.csswg.org/css-pseudo/#treelike).

This would allow `::tooltip` to inherit properties, like `font-family`,
from the rest of the page, making it easier for authors to create
a consistent look and feel across all their UI.

### What user interaction triggers the built-in tooltip?

Currently, browsers show `title`-based tooltips when a user hovers
over the element associated with the tooltip.

> No major browser engines appears to currently support touch interactions

Reviewer 4: ```suggestion
No major browser engine appears to currently support touch interactions
```

Authors may also want to be able to customize the interaction(s) that
trigger built-in tooltips using the `::tooltip` pseudo element. This
functionality is being considered as a [future addition](#future-ideas)
to this proposal.

### Dependencies on non-stable features

This proposal makes use of [CSS Anchor Positioning](
https://drafts.csswg.org/css-anchor-position-1/), to define the default
positioning of the base `::tooltip` styles. This feature is not yet
baseline, but it is part of [Interop 2025](
https://github.com/web-platform-tests/interop/blob/main/2025/README.md#css-anchor-positioning).

There is also a thought on whether `::tooltip` should be defined to be
based on `interesttarget`, given there is an overlap between the two
features. However, this is something that should likely be left up to the
UA on the technology involved in rendering their tooltips, as long as
the base set of styles and specified interactions are met.

## Key scenarios

### Scenario 1: Opt-into the new base `::tooltip` styles

In its most base form, an author could opt-into the new `::tooltip` base
styles without adjusting any of its styles in order to get a consistent basic

> styling for tooltips across browsres. An author could accomplish this with

Reviewer 4: ```suggestion
styling for tooltips across browsers. An author could accomplish this with
```

Author: Done

of the Iframe window. This, however, would be consistent with the behavior
authors get with custom JS tooltips already, so applying the same restriction
here likely wouldn’t be a major downgrade for authors.

## Privacy Considerations

There are no known privacy considerations to take into account as part of
this proposal.

## Open questions

- [What styles should be allowed within the `::tooltip` pseudo
element?](https://github.com/w3c/csswg-drafts/issues/9447)
- Should `::tooltip` be based on `interesttarget` or should that
detail be left up to the UA?
- Should we consider another method for triggering `::tooltip` and its
base styles than the `appearance` property?
- Is `content` the best mechanism for setting the tooltip text from
`title` within the `::tooltip` psuedo element?
- Should authors be allowed to change the `content` of the `::tooltip`?
- Should authors be able to change the positioning of a `::tooltip` or
should this be magic left up to the UA?
- What are the right default UA styles for `::tooltip`?
- Does it make sense for `::tooltip` to be a [Tree-Abiding
Pseudo-element](https://drafts.csswg.org/css-pseudo/#treelike)?

> - How should `::tootltip` interact with `interesttarget` and any other

Reviewer 4: ```suggestion
- How should `::tooltip` interact with `interesttarget` and any other
```
